### PR TITLE
Implement Makie extension

### LINKED ISSUE #3178: Adding Makie extension
I was reading about the extension feature in Julia v1.9

https://pkgdocs.julialang.org/v1/creating-packages/#Conditional-loading-of-code-in-packages-(Extensions)

and it reads to me that it's ideal to add for plotting methods that step onto Makie functionality. This way the code won't load unless one loads GLMakie/CairoMakie in the environment. What do others think?

I have put together a few methods for plotting Oceananigans fields at https://github.com/navidcy/Imaginocean.jl. I could move those into an extension in Oceananigans.

Ideally I would like these things to live outside Oceananigans repo to minimise the burden of maintenance. But from what I understand for how extensions work they have to live in this repo? Or in the Makie repo which makes even less sense? Am I right?

x-ref: https://github.com/navidcy/Imaginocean.jl/issues/2

### PR DESCRIPTION
**Pull request by glwagner**
This PR implements an extension to Oceananigans with `convert_arguments` for Makie plots.

For example:

```julia
julia> using Oceananigans, GLMakie
Precompiling Oceananigans
  7 dependencies successfully precompiled in 18 seconds. 157 already precompiled.
[ Info: Precompiling GLMakie [e9467ef8-e4e7-5192-8a1a-b1aee30e663a]
[ Info: Precompiling OceananigansMakieExt [8b7e02c2-18e1-5ade-af7b-cfb5875075c8]
[ Info: Precompiling ChainRulesCoreExt [eae2faf6-b232-58cb-a410-7764fda2830c]

julia> grid = RectilinearGrid(size=(10, 10, 10), x=(0, 1), y=(0, 1), z=(0, 1))
10×10×10 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── Periodic x ∈ [0.0, 1.0) regularly spaced with Δx=0.1
├── Periodic y ∈ [0.0, 1.0) regularly spaced with Δy=0.1
└── Bounded  z ∈ [0.0, 1.0] regularly spaced with Δz=0.1

julia> c = CenterField(grid)
10×10×10 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 10×10×10 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 16×16×16 OffsetArray(::Array{Float64, 3}, -2:13, -2:13, -2:13) with eltype Float64 with indices -2:13×-2:13×-2:13
    └── max=0.0, min=0.0, mean=0.0

julia> set!(c, (x, y, z) -> rand())
10×10×10 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 10×10×10 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 16×16×16 OffsetArray(::Array{Float64, 3}, -2:13, -2:13, -2:13) with eltype Float64 with indices -2:13×-2:13×-2:13
    └── max=0.998531, min=0.000368158, mean=0.516368

julia> heatmap(view(c, :, 2, :))
```

image

I edited a few of the examples and enabled push previews, so we can take a look at the docs when they are finished.

The next level is to automatically label axes, but I think we should save that for a future PR.

Closes #3178

### CODE REVIEW

File: ext/OceananigansMakieExt.jl [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        return 1
+    else
+        return Colon()
+    end
+end
+
+function convert_arguments(pl::Type{<:AbstractPlot}, f::Field)
+
+    # Deduce dimensionality
+    Nx, Ny, Nz = size(f)
```
Comment (navidcy) [inline ext/OceananigansMakieExt.jl:20]: so are you implying that f should be a 2D slice?
if so, how would this work... size would return a 2-tuple...

Comment (glwagner) [reply ext/OceananigansMakieExt.jl:20]: The code I wrote works --- its not a concept! `size` is always a 3-tuple:

```julia
julia> size(c)
(10, 10, 10)

julia> size(view(c, :, 2, :))
(10, 1, 10)
```

Comment (glwagner) [reply ext/OceananigansMakieExt.jl:20]:
```julia
julia> r = Field{Center, Center, Nothing}(grid)
si10×10×1 Field{Center, Center, Nothing} reduced over dims = (3,) on RectilinearGrid on CPU
├── grid: 10×10×10 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: Nothing, top: Nothing, immersed: ZeroFlux
└── data: 16×16×1 OffsetArray(::Array{Float64, 3}, -2:13, -2:13, 1:1) with eltype Float64 with indices -2:13×-2:13×1:1
    └── max=0.0, min=0.0, mean=0.0

julia> size(r)
(10, 10, 1)

julia> grid = RectilinearGrid(size=(10, 10), x=(0, 1), y=(0, 1), topology=(Periodic, Periodic, Flat))
10×10×1 RectilinearGrid{Float64, Periodic, Periodic, Flat} on CPU with 3×3×0 halo
├── Periodic x ∈ [0.0, 1.0) regularly spaced with Δx=0.1
├── Periodic y ∈ [0.0, 1.0) regularly spaced with Δy=0.1
└── Flat z

julia> c = CenterField(grid)
si10×10×1 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 10×10×1 RectilinearGrid{Float64, Periodic, Periodic, Flat} on CPU with 3×3×0 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: Nothing, top: Nothing, immersed: ZeroFlux
└── data: 16×16×1 OffsetArray(::Array{Float64, 3}, -2:13, -2:13, 1:1) with eltype Float64 with indices -2:13×-2:13×1:1
    └── max=0.0, min=0.0, mean=0.0

julia> size(c)
(10, 10, 1)

Comment (navidcy) [reply ext/OceananigansMakieExt.jl:20]: Gotcha! I was thinking of 

```Julia
julia> size(interior(c, :, :, 1))
(10, 10)
```

**Comment by glwagner** [comment]
@navidcy is there any more functionality from [`Imaginocean`](https://github.com/navidcy/Imaginocean.jl) that would make sense to sneak into this PR?

File: ext/OceananigansMakieExt.jl (hunk 1 above)
Comment (glwagner) [reply ext/OceananigansMakieExt.jl:20]: Right, `interior` returns a `SubArray` but this is just for fields.

Comment (navidcy) [reply ext/OceananigansMakieExt.jl:20]: My bad though since `interior(c, :, :, 1)` is _not_ a field!

**Comment by navidcy** [comment]
Let's start with this! How does it work for a lat-lon field?
Does it work for, e.g., XFaceFields?

File: examples/two_dimensional_turbulence.jl [hunk 2]
```diff
@@ -135,7 +135,7 @@ nothing #hide
 
 # and animate the vorticity and fluid speed.
 
-using CairoMakie
+using GLMakie #CairoMakie
```
Comment (navidcy) [inline examples/two_dimensional_turbulence.jl:138]: remove GLMakie

File: examples/baroclinic_adjustment.jl [hunk 3]
```diff
@@ -80,8 +80,9 @@ x, y, z = 1e-3 .* nodes(grid, (Center(), Center(), Center()))
 
 b = model.tracers.b
 
-fig, ax, hm = heatmap(y, z, interior(b)[1, :, :],
-                      colormap=:deep,
+#fig, ax, hm = heatmap(y, z, interior(b)[1, :, :],
```
Comment (navidcy) [inline examples/baroclinic_adjustment.jl:83]:
```suggestion
```

File: examples/baroclinic_adjustment.jl [hunk 4]
```diff
@@ -331,6 +332,7 @@ B = @lift interior(B_timeseries[$n], 1, :, :)
 
 # and then build our plot:
 
+#hm = heatmap!(axb, xb, yb, b_top, colorrange=(0, Δb), colormap=:thermal)
```
Comment (navidcy) [inline examples/baroclinic_adjustment.jl:335]:
```suggestion
```

File: examples/two_dimensional_turbulence.jl (hunk 2 above)
Comment (navidcy) [inline examples/two_dimensional_turbulence.jl:138]:
```suggestion
using CairoMakie
```

**Review by navidcy — APPROVED** [review]
great!

if docs build successfully merge!

**Comment by navidcy** [comment]
```Julia
julia> using Oceananigans, GLMakie
[ Info: Oceananigans will use 8 threads
[ Info: Precompiling TaylorSeriesIAExt [ed7ef945-33a4-511e-97fe-2b89c7a130ca]
ERROR: LoadError: UndefVarError: `IntervalBox` not defined
Stacktrace:
 [1] top-level scope
   @ ~/.julia/packages/TaylorSeries/2qRvJ/ext/TaylorSeriesIAExt.jl:182
 [2] include
   @ ./Base.jl:495 [inlined]
 [3] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
   @ Base ./loading.jl:2222
 [4] top-level scope
   @ stdin:3
in expression starting at /Users/navid/.julia/packages/TaylorSeries/2qRvJ/ext/TaylorSeriesIAExt.jl:1
in expression starting at stdin:3
┌ Error: Error during loading of extension TaylorSeriesIAExt of TaylorSeries, use `Base.retry_load_extensions()` to retry.
│   exception =
│    1-element ExceptionStack:
│    Failed to precompile TaylorSeriesIAExt [ed7ef945-33a4-511e-97fe-2b89c7a130ca] to "/Users/navid/.julia/compiled/v1.10/TaylorSeriesIAExt/jl_TNauRw".
│    Stacktrace:
│      [1] error(s::String)
│        @ Base ./error.jl:35
│      [2] compilecache(pkg::Base.PkgId, path::String, internal_stderr::IO, internal_stdout::IO, keep_loaded_modules::Bool)
│        @ Base ./loading.jl:2468
│      [3] compilecache
│        @ ./loading.jl:2340 [inlined]
│      [4] (::Base.var"#968#969"{Base.PkgId})()
│        @ Base ./loading.jl:1974
│      [5] mkpidlock(f::Base.var"#968#969"{Base.PkgId}, at::String, pid::Int32; kwopts::@Kwargs{stale_age::Int64, wait::Bool})
│        @ FileWatching.Pidfile ~/julia-1.10/usr/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:93
│      [6] #mkpidlock#6
│        @ ~/julia-1.10/usr/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:88 [inlined]
│      [7] trymkpidlock(::Function, ::Vararg{Any}; kwargs::@Kwargs{stale_age::Int64})
│        @ FileWatching.Pidfile ~/julia-1.10/usr/share/julia/stdlib/v1.10/FileWatching/src/pidfile.jl:111
│      [8] #invokelatest#2
│        @ ./essentials.jl:894 [inlined]
│      [9] invokelatest
│        @ ./essentials.jl:889 [inlined]
│     [10] maybe_cachefile_lock(f::Base.var"#968#969"{Base.PkgId}, pkg::Base.PkgId, srcpath::String; stale_age::Int64)
│        @ Base ./loading.jl:2983
│     [11] maybe_cachefile_lock
│        @ ./loading.jl:2980 [inlined]
│     [12] _require(pkg::Base.PkgId, env::Nothing)
│        @ Base ./loading.jl:1970
│     [13] __require_prelocked(uuidkey::Base.PkgId, env::Nothing)
│        @ Base ./loading.jl:1812
│     [14] #invoke_in_world#3
│        @ ./essentials.jl:926 [inlined]
│     [15] invoke_in_world
│        @ ./essentials.jl:923 [inlined]
│     [16] _require_prelocked
│        @ ./loading.jl:1803 [inlined]
│     [17] _require_prelocked
│        @ ./loading.jl:1802 [inlined]
│     [18] run_extension_callbacks(extid::Base.ExtensionId)
│        @ Base ./loading.jl:1295
│     [19] run_extension_callbacks(pkgid::Base.PkgId)
│        @ Base ./loading.jl:1330
│     [20] run_package_callbacks(modkey::Base.PkgId)
│        @ Base ./loading.jl:1164
│     [21] _tryrequire_from_serialized(modkey::Base.PkgId, path::String, ocachepath::String, sourcepath::String, depmods::Vector{Any})
│        @ Base ./loading.jl:1487
│     [22] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String, build_id::UInt128)
│        @ Base ./loading.jl:1574
│     [23] _require(pkg::Base.PkgId, env::String)
│        @ Base ./loading.jl:1938
│     [24] __require_prelocked(uuidkey::Base.PkgId, env::String)
│        @ Base ./loading.jl:1812
│     [25] #invoke_in_world#3
│        @ ./essentials.jl:926 [inlined]
│     [26] invoke_in_world
│        @ ./essentials.jl:923 [inlined]
│     [27] _require_prelocked(uuidkey::Base.PkgId, env::String)
│        @ Base ./loading.jl:1803
│     [28] macro expansion
│        @ ./loading.jl:1790 [inlined]
│     [29] macro expansion
│        @ ./lock.jl:267 [inlined]
│     [30] __require(into::Module, mod::Symbol)
│        @ Base ./loading.jl:1753
│     [31] #invoke_in_world#3
│        @ ./essentials.jl:926 [inlined]
│     [32] invoke_in_world
│        @ ./essentials.jl:923 [inlined]
│     [33] require(into::Module, mod::Symbol)
│        @ Base ./loading.jl:1746
│     [34] eval
│        @ ./boot.jl:385 [inlined]
│     [35] eval_user_input(ast::Any, backend::REPL.REPLBackend, mod::Module)
│        @ REPL ~/julia-1.10/usr/share/julia/stdlib/v1.10/REPL/src/REPL.jl:150
│     [36] repl_backend_loop(backend::REPL.REPLBackend, get_module::Function)
│        @ REPL ~/julia-1.10/usr/share/julia/stdlib/v1.10/REPL/src/REPL.jl:246
│     [37] start_repl_backend(backend::REPL.REPLBackend, consumer::Any; get_module::Function)
│        @ REPL ~/julia-1.10/usr/share/julia/stdlib/v1.10/REPL/src/REPL.jl:231
│     [38] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool, backend::Any)
│        @ REPL ~/julia-1.10/usr/share/julia/stdlib/v1.10/REPL/src/REPL.jl:389
│     [39] run_repl(repl::REPL.AbstractREPL, consumer::Any)
│        @ REPL ~/julia-1.10/usr/share/julia/stdlib/v1.10/REPL/src/REPL.jl:375
│     [40] (::Base.var"#1013#1015"{Bool, Bool, Bool})(REPL::Module)
│        @ Base ./client.jl:432
│     [41] #invokelatest#2
│        @ ./essentials.jl:892 [inlined]
│     [42] invokelatest
│        @ ./essentials.jl:889 [inlined]
│     [43] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
│        @ Base ./client.jl:416
│     [44] exec_options(opts::Base.JLOptions)
│        @ Base ./client.jl:333
│     [45] _start()
│        @ Base ./client.jl:552
└ @ Base loading.jl:1301
```

perhaps some compat entry is needed? I get this w GLMakie v0.10.5 in my main Julia env....

**Comment by glwagner** [comment]
huh, not sure

**Comment by navidcy** [comment]
what GLMakie are you using?

**Comment by glwagner** [comment]
> what GLMakie are you using?

```
(test) pkg> st
Status `~/Projects/test/Project.toml`
  [e9467ef8] GLMakie v0.10.5
  [9e8cae18] Oceananigans v0.91.5 `~/Projects/Oceananigans.jl`
```

**Comment by glwagner** [comment]
> Let's start with this! How does it work for a lat-lon field? Does it work for, e.g., XFaceFields?

it should work for a lat lon field I think, but it won't work for `OrthogonalSphericalShellGrid` without the extra stuff from `Imaginocean`, I think

**Comment by glwagner** [comment]
Preview: https://clima.github.io/OceananigansDocumentation/previews/PR3686

**Comment by navidcy** [comment]
> Preview: https://clima.github.io/OceananigansDocumentation/previews/PR3686

You are anticipating! Nice!

**Comment by navidcy** [comment]
Docs failed. I suggest try building locally first! Comment out all examples but one.

**Comment by navidcy** [comment]
See https://clima.github.io/OceananigansDocumentation/stable/contributing/#Documentation

**Comment by glwagner** [comment]
Isn't it easier to run the example directly?

**Comment by navidcy** [comment]
If you wanna ensure that you are doing exactly what the docs built will be doing then do the above.
Otherwise, yes just running the example is easier. I thought you'd run it already :)

**Comment by glwagner** [comment]
> If you wanna ensure that you are doing exactly what the docs built will be doing then do the above. Otherwise, yes just running the example is easier. I thought you'd run it already so I was trying to find the culprit :)

I started by updating all the examples for the new syntax that I thought would work, and then went on to other stuff to let the CI do its thing. Now that I see one of them failed I went back and focused on that one specifically. I think its an ok workflow since regardless I would have to wait for compilation / setup, so its not a bad strategy to let CI find the errors for you, at least at first

**Comment by navidcy** [comment]
> I started by updating all the examples for the new syntax that I thought would work, and then went on to other stuff to let the CI do its thing. Now that I see one of them failed I went back and focused on that one specifically. I think its an ok workflow since regardless I would have to wait for compilation / setup, so its not a bad strategy to let CI find the errors for you, at least at first

It is, if it works for you!
Sometimes I personally have a larger adjustment time to get back to a problem when I move to a different one so the workflow you seem to follow doesn't work for me that well.

Merge?

**Comment by glwagner** [comment]
> > I started by updating all the examples for the new syntax that I thought would work, and then went on to other stuff to let the CI do its thing. Now that I see one of them failed I went back and focused on that one specifically. I think its an ok workflow since regardless I would have to wait for compilation / setup, so its not a bad strategy to let CI find the errors for you, at least at first
> 
> It is, if it works for you! Sometimes I personally have a larger adjustment time to get back to a problem when I move to a different one so the workflow you seem to follow doesn't work for me that well.
> 
> Merge?

True... I'm working on 30 things at once usually...

**Pull request merged**